src/main/java/io/lettuce/authx/TokenBasedRedisCredentialsProvider.java

```diff
@@ -0,0 +1,108 @@
+package io.lettuce.authx;
+
+import io.lettuce.core.RedisCredentials;
+import io.lettuce.core.RedisCredentialsProvider;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Sinks;
+import redis.clients.authentication.core.Token;
+import redis.clients.authentication.core.TokenAuthConfig;
+import redis.clients.authentication.core.TokenListener;
+import redis.clients.authentication.core.TokenManager;
+
+public class TokenBasedRedisCredentialsProvider implements RedisCredentialsProvider, AutoCloseable {
+
+    private final TokenManager tokenManager;
+
+    private final Sinks.Many<RedisCredentials> credentialsSink = Sinks.many().replay().latest();
+
+    public TokenBasedRedisCredentialsProvider(TokenAuthConfig tokenAuthConfig) {
+        this(new TokenManager(tokenAuthConfig.getIdentityProviderConfig().getProvider(),
+                tokenAuthConfig.getTokenManagerConfig()));
+
+    }
+
+    public TokenBasedRedisCredentialsProvider(TokenManager tokenManager) {
+        this.tokenManager = tokenManager;
+        initializeTokenManager();
+    }
+
+    /**
+     * Initialize the TokenManager and subscribe to token renewal events.
+     */
+    private void initializeTokenManager() {
+        TokenListener listener = new TokenListener() {
+
+            @Override
+            public void onTokenRenewed(Token token) {
+                try {
+                    String username = token.getUser();
+                    char[] pass = token.getValue().toCharArray();
+                    RedisCredentials credentials = RedisCredentials.just(username, pass);
+                    credentialsSink.tryEmitNext(credentials);
+                } catch (Exception e) {
```

Comment:
As discussed - let's try to remove this for now.

Background : multiple threads calling `tryEmitNext` could cause and exception, as it is not thread safe. This is a [serious shortcoming of the Flux Sinks](https://stackoverflow.com/questions/65029619/how-to-call-sinks-manyt-tryemitnext-from-multiple-threads) that has no easy workaround.

Since we are not sure there would ever be such a case we will let this exception bubble up, but we will not complete the flux with error (as this seems more drastic and will have the side effect of not consuming any more updates from the token manager)

Reply:
Agree. Will update the review.